### PR TITLE
Fix sobel

### PR DESCRIPTION
**Pull request by pranavb-ca**

### CODE REVIEW

File: src/HexagonOptimize.cpp [hunk 1]
```diff
@@ -214,12 +220,17 @@ Expr apply_patterns(Expr x, const std::vector<Pattern> &patterns, IRMutator *op_
...
+                if ((p.flags & Pattern::LosslessCastOp1) ||
+                    (p.flags & Pattern::LosslessCastOp0)) {
                     internal_assert(matches.size() == 2);
-                    Type t = Type(matches[1].type().code(), matches[1].type().bits()/2, 1);
-                    Expr b = lossless_cast(t, matches[1]);
+                    // LosslessCastOpx is a way of knowing that operand x is a broadcast.
+                    // Therefore, both *_should_not_* be set in the pattern flags because
+                    // that'll mean we are operating on two broadcasts.
+                    int op_num = (p.flags & Pattern::LosslessCastOp0) ? 0 : 1;
+                    Type t = Type(matches[op_num].type().code(), matches[op_num].type().bits()/2, 1);
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: Type has with_* methods that make this easier (IMO):

    Type t = matches[op_num].type();
    t = t.with_bits(t.bits() / 2).with_lanes(1);

**Comment by dsharletg** [comment]
Thanks, this looks great!

File: src/HexagonOptimize.cpp (hunk 1 above)
Comment (pranavb-ca) [inline src/HexagonOptimize.cpp]: On it.

File: src/CodeGen_Hexagon.cpp [hunk 2]
```diff
@@ -371,9 +370,6 @@ llvm::Function *CodeGen_Hexagon::define_hvx_intrinsic(Intrinsic::ID id, Type ret
     llvm::Function *intrin = Intrinsic::getDeclaration(module.get(), id);
     return define_hvx_intrinsic(intrin, ret_ty, name, arg_types, broadcast_scalar);
 }
-llvm::Value *CodeGen_Hexagon::shl_or(llvm::Value *v, unsigned u) {
```
Comment (dsharletg) [inline src/CodeGen_Hexagon.cpp]: Remember to remove this from src/CodeGen_Hexagon.h as well.

File: src/CodeGen_Hexagon.cpp [hunk 3]
```diff
@@ -371,9 +370,6 @@ llvm::Function *CodeGen_Hexagon::define_hvx_intrinsic(Intrinsic::ID id, Type ret
...
-    return builder->CreateOr(builder->CreateShl(v, u), v);
- }
 llvm::Function *CodeGen_Hexagon::define_hvx_intrinsic(llvm::Function *intrin, Type ret_ty, const std::string &name,
                                                       const std::vector<Type> &arg_types, bool broadcast_scalar) {
```
Comment (dsharletg) [inline src/CodeGen_Hexagon.cpp]: Sorry, one more thing: can you rename broadcast_scalar to broadcast_scalar_to_word? Or something like that. It's ambiguous whether that means broadcast the scalar all the way up to a vector, or just to a word like it's doing.

Comment (pranavb-ca) [inline src/CodeGen_Hexagon.cpp]: Makes sense.

File: src/HexagonOptimize.cpp [hunk 4]
```diff
@@ -186,10 +187,15 @@ std::vector<Pattern> muls = {
...
-    { "halide.hexagon.mpy.vub.ub", u16(wild_u8x)*bc(wild_u16), Pattern::InterleaveResult | Pattern::LosslessCast },
-    { "halide.hexagon.mpy.vub.b", i16(wild_u8x)*bc(wild_i16), Pattern::InterleaveResult | Pattern::LosslessCast },
-    { "halide.hexagon.mpy.vuh.uh", u32(wild_u16x)*bc(wild_u32), Pattern::InterleaveResult | Pattern::LosslessCast },
-    { "halide.hexagon.mpy.vh.h",  i32(wild_i16x)*bc(wild_i32), Pattern::InterleaveResult | Pattern::LosslessCast },
+    { "halide.hexagon.mpy.vub.ub", u16(wild_u8x)*bc(wild_u16), Pattern::InterleaveResult | Pattern::LosslessCastOp1 },
+    { "halide.hexagon.mpy.vub.b", i16(wild_u8x)*bc(wild_i16), Pattern::InterleaveResult | Pattern::LosslessCastOp1 },
+    { "halide.hexagon.mpy.vuh.uh", u32(wild_u16x)*bc(wild_u32), Pattern::InterleaveResult | Pattern::LosslessCastOp1 },
+    { "halide.hexagon.mpy.vh.h",  i32(wild_i16x)*bc(wild_i32), Pattern::InterleaveResult | Pattern::LosslessCastOp1 },
+    // Commute the four above.
+    { "halide.hexagon.mpy.vub.ub", bc(wild_u16)*u16(wild_u8x), Pattern::InterleaveResult | Pattern::LosslessCastOp0 },
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: Don't these need Pattern::SwapOperands?

File: src/HexagonOptimize.cpp [hunk 5]
```diff
@@ -214,12 +220,17 @@ Expr apply_patterns(Expr x, const std::vector<Pattern> &patterns, IRMutator *op_
                 if (p.flags & Pattern::SwapOperands) {
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: I think the SwapOperands handler should go last (after the LosslessCastOp handler). I was thinking of SwapOperands as swapping prior to substitution, so it wouldn't affect the operand index of LosslessCastOp

Comment (pranavb-ca) [inline src/HexagonOptimize.cpp]: You are right. Done.

**Comment by dsharletg** [comment]
Thanks for addressing all the comments, sorry they trickled in one at a time as I noticed them :)

**Pull request merged**